### ISSUE TITLE
SDL Example: Assertion io.DeltaTime >= 0 in examples when FPS is very high

### ISSUE THREAD

**Issue by Wizzard033** [ocornut/imgui#996]
The following line in imgui_impl_sdl_gl3.cpp fails very often

    io.DeltaTime = g_Time > 0.0 ? (float)(current_time - g_Time) : (float)(1.0f / 60.0f);

It fails due to the assertion io.DeltaTime >= 0 which isn't true here
I see this same line in a lot of the examples, and I don't know how it's managed to get this far
Maybe it is just a MinGW thing that causes this to be negative at very high framerates (5000+)

My fix I did follows

    io.DeltaTime = g_Time > 0.0 ? std::max(0.f, (float)(current_time - g_Time)) : (float)(1.0f / 60.0f);

**Comment by ocornut** [comment]
Your suggested fix is incorrect at it will create spikes of delta time. Those times are used for various time measurement (mostly the double click and various repeat speed).

You should investigate why you are getting a negative DeltaTime in the first place.

**Comment by Wizzard033** [comment]
I just compile the example and get this issue. I believe its due to hardware and/or MinGW.
It only happens when the FPS is greater than about 5000.

**Comment by Wizzard033** [comment]
Okay! I tested a new way to fix it and it worked. Let me know if it's better.

    // Setup time step
    Uint32 current_time = SDL_GetTicks();
    io.DeltaTime = g_Time > 0 ? (float)((current_time - g_Time) / 1000.0f) : (float)(1.0f / 60.0f);
    g_Time = current_time;

I had to change the type of g_Time to Uint32 near the top of imgui_impl_sdl_gl3.cpp too.

**Comment by Wizzard033** [comment]
I'd also like to mention that the example is **constantly** spamming the assertion when ran.

**Comment by ocornut** [comment]
Could you log, say, the 1000 first values returned by SDL_GetTicks() in that case and then dumping them out?

Something like:
```
static Uint32 buf[1000];
buf_n = 0;
Uint32 current_time = SDL_GetTicks();
if (buf_n < 1000) 
  buf[buf_n] = current_time;
buf_n++;
if (buf_n == 1000)
{
  FILE* f = fopen("log.txt", "wt");
  for (int n = 0; n < 1000; n++) 
    fprintf(f, "[%04d] %u\n", n, buf[n]);
  fclose(f);
}
```

**Comment by Wizzard033** [comment]
[log.txt](https://github.com/ocornut/imgui/files/737265/log.txt)

**Comment by Wizzard033** [comment]
Here's another log I did using the following code

    static double buf[1000];
    static int buf_n = 0;
    double time = SDL_GetTicks() / 1000.0;
    if (buf_n < 1000)
        buf[buf_n] = time;
    buf_n++;
    if (buf_n == 1000) {
        FILE* f = fopen("log.txt", "wt");
        for (int n = 1; n < 1000; n++) {
            float result = (float)(buf[n] - buf[n - 1]);
            fprintf(f, "[%04d] prev = %04f, curr = %04f, (curr - prev) = %04f, negative = %s\n", n, buf[n - 1], buf[n], result, result < 0 ? "true" : "false");
        }
        fclose(f);
    }


[log-exception.txt](https://github.com/ocornut/imgui/files/737280/log-exception.txt)

During the log, I had to click "Ignore" on many, many exceptions...

**Comment by Wizzard033** [comment]
One more test I did that makes sure the log even more representative
It does this by interspersing with the actual time code and using an assert that never failed

    // Setup time step
    static double buf[1000];
    static int buf_n = 0;
    Uint32	time = SDL_GetTicks();
    double current_time = time / 1000.0;
    if (buf_n < 1000)
        buf[buf_n] = current_time;
    io.DeltaTime = g_Time > 0.0 ? (float)(current_time - g_Time) : (float)(1.0f / 60.0f);
    assert((buf_n == 0) || (g_Time == buf[buf_n - 1]));
    g_Time = current_time;
    buf_n++;
    if (buf_n == 1000) {
        FILE* f = fopen("log.txt", "wt");
        for (int n = 1; n < 1000; n++) {
            float result = (float)(buf[n] - buf[n - 1]);
            fprintf(f, "[%04d] prev = %04f, curr = %04f, (curr - prev) = %04f, negative = %s\n", n, buf[n - 1], buf[n], result, result < 0 ? "true" : "false");
        }
        fclose(f);
    }

[log-exception2.txt](https://github.com/ocornut/imgui/files/737292/log-exception2.txt)

Once again I had to click "Ignore" on many, many exceptions, but all for io.DeltaTime.

**Comment by Wizzard033** [comment]
The assert actually failed in the previous test. I forgot to compile in debug mode.

**Comment by Wizzard033** [comment]
It seems as if my fix _would_ work if the library allowed io.DeltaTime = 0 (see the referenced issue)

**Comment by ocornut** [comment]
I haven't have time to look at it. 

Allowing io.DeltaTime = 0.0f makes little sense and even if some workaround would be possible in term of handling buttons, other things would break.

Make sure your app doesn't give DeltaTime 0.0 (the sdl2 example code probably needs fixing as you suggested above, but passing 0.0f isn't a correct fix).

**Comment by ocornut** [comment]
Replying here instead of the other closed thread.

> It's incorrect to use a millisecond precision timer and have greater than 1000 FPS with your library?

It's incorrect to give a zero DeltaTime.

If your application wants to deal with 1000+ FPS framerate then surely your application shouldn't be using a millisecond precision timer? If you can get a precise enough value at least uses 1.0f/FPS instead of zero..

**Comment by Wizzard033** [comment]
Okay. That makes sense I guess. I'll go ahead and limit FPS at 1000 for my application. Thanks!
I think there should be some note made of this in the documentation so others don't waste hours like me.
The original issue still applies, and fixing the example is more involved as it needs to have it's FPS capped.

**Comment by ocornut** [comment]
Yes there's probably an issue with that SDL2 example, I will look at it.
You don't need to cap your app to 1000 PFS to instead of passing 0.0f as a fallback pass 1.0f/FPS as a vague estimation of time passing.

**Comment by Wizzard033** [comment]
Good suggestion. Thanks!

**Comment by paulsapps** [comment]
If you are using C++11 then maybe use this: http://stackoverflow.com/questions/26440860/sdl-getticks-accuracy-below-the-millisecond-level

**Comment by Wizzard033** [comment]
Does ImGui use cumulative time for anything? If so, nanoseconds is probably too much for a float.
This is why I didn't go that route, but without knowing how ImGui works, I can't tell.

**Comment by paulsapps** [comment]
If you up convert the units it still won't be zero..

**Comment by AlfonsoJLuna** [comment]
SDL_GetPerformanceCounter() / SDL_GetPerformanceFrequency() can be used for a more precise timer.
This could be a solution:

    // Setup time step
    double current_time = SDL_GetPerformanceCounter() / SDL_GetPerformanceFrequency();
    io.DeltaTime = g_Time > 0.0 ? (float)(current_time - g_Time) : (float)(1.0f/60.0f);
    g_Time = current_time;

**Comment by ocornut** [comment]
@Wizzard033 I finally got around to fix this (using the performance counter as suggested by @AlfonsoJLuna). 

My apologies Brandon for taking so long to tackle such a simple problem. Today I was testing Navigation branch input and the problem was very noticeable due to how the Nav branch handle its inputs (it's currently not tolerant to `io.DeltaTime` being 0.0f)

Note that I also changed the SDL example to enable vsync so they'll probably be capped to 60 hz etc. by default but I tested it for high framerates.
